FWCore/Concurrency/interface/chain_first.h

```diff
@@ -0,0 +1,273 @@
+       argument will be non-nullptr. In that case, the exception will NOT be automatically propagated to the WaitingTaskHolder. In addition,
+       if the functor itself throws an exception, it is up to the functor to handle the exception.
+       */
+      template <typename O>
+      [[nodiscard]] constexpr auto then(O&& iO) {
+        return _then(std::forward<O>(iO), std::move(*this));
+      }
+
+      template <typename O>
+      [[nodiscard]] constexpr auto ifThen(bool iCondition, O&& iO) {
+        return WaitingTaskChain<Conditional<AutoExceptionHandler<O>>, Conditional<U>, T...>(
+            iCondition, AutoExceptionHandler<O>(std::forward<O>(iO)), std::move(*this));
+      }
+
+      [[nodiscard]] WaitingTaskHolder lastTask(WaitingTaskHolder iV) {
+        if (condition_) {
+          return l_.lastTask(WaitingTaskHolder(
+              *iV.group(),
+              make_waiting_task([f = std::move(f_), v = std::move(iV)](std::exception_ptr const* iPtr) mutable {
+                f(iPtr, std::move(v));
+              })));
+        }
+        return l_.lastTask(iV);
+      }
+
+      void runLast(WaitingTaskHolder iV) {
+        if (condition_) {
+          l_.runLast(WaitingTaskHolder(
+              *iV.group(),
+              make_waiting_task([f = std::move(f_), v = std::move(iV)](std::exception_ptr const* iPtr) mutable {
+                f(iPtr, std::move(v));
+              })));
+        } else {
+          l_.runLast(iV);
+        }
+      }
+
+    private:
+      WaitingTaskChain<T...> l_;
+      U f_;
+      bool condition_;
+    };
+
+  }  // namespace waiting_task::detail
+  namespace waiting_task::chain {
+
+    /** Sets the first task to be run as part of the task chain. The functor is expected to take either
+   a single argument of type edm::WairingTaskHolder or two arguments of type std::exception_ptr const* and WaitingTaskHolder. In the latter case, the pointer is only non-null if a previous task in the chain threw an exception.
+   */
+    template <typename F>
+    auto first(F&& iF) {
+      using namespace detail;
+      if constexpr (has_exception_handling<F>::value) {
+        return WaitingTaskChain<F>(std::forward<F>(iF));
+      } else {
+        return WaitingTaskChain<AutoExceptionHandler<F>>(AutoExceptionHandler<F>(std::forward<F>(iF)));
+      }
+    }
+
+    /**Creates a functor adaptor which assembled two different functors into one. To use, one calls the constructor immediately followed by the else_ method. The created functor has the following behavior:
+   If a previous task had an exception, only the first functor given to the constructor, iE, will be run and passed the std::exception_ptr const. If there
+   was no exception, then only the functor passed to else_, iF, will be run. If iF has an exception, it will be automatically propagated to the edm::WaitingTaskHolder. */
+    template <typename E>
+    struct IfException {
```

Comment:
Functionally the `IfException` seems to be used like the `first()`. Even if this would technically violate our naming rules, I'm wondering about
```suggestion
    struct ifException {
```
Then all the pieces of the chaining would start with lower case letter.

Reply:
I could add a function `ifException` which just calls the relevant constructor. That way we do not need to violate our naming conventions.